inject/__init__.py

```diff
@@ -382,11 +246,142 @@ def __call__(self, func):
         @wraps(func)
         def injection_wrapper(*args, **kwargs):
 
-            provided_params = frozenset(arg_names[:len(args)]) | frozenset(kwargs.keys())
+            provided_params = frozenset(
+                arg_names[:len(args)]) | frozenset(kwargs.keys())
             for param, cls in params_to_provide.items():
                 if param not in provided_params:
                     kwargs[param] = instance(cls)
-
             return func(*args, **kwargs)
-
         return injection_wrapper
+
+
+def configure(config=None, bind_in_runtime=True) -> Injector:
+    """Create an injector with a callable config or raise an exception when already configured."""
+    global _INJECTOR
+
+    with _INJECTOR_LOCK:
+        if _INJECTOR:
+            raise InjectorException('Injector is already configured')
+
+        _INJECTOR = Injector(config, bind_in_runtime=bind_in_runtime)
+        logger.debug('Created and configured an injector, config=%s', config)
+        return _INJECTOR
+
+
+def configure_once(config=None, bind_in_runtime=True) -> Injector:
+    # type: (Optional[BinderCallable], bool) -> Injector
+    """Create an injector with a callable config if not present, otherwise, do nothing."""
+    with _INJECTOR_LOCK:
+        if _INJECTOR:
+            return _INJECTOR
+
+        return configure(config, bind_in_runtime=bind_in_runtime)
+
+
+def clear_and_configure(config: Optional[BinderCallable] = None, bind_in_runtime: bool = True) -> Injector:
+    # type: (, bool) -> Injector
+    """Clear an existing injector and create another one with a callable config."""
+    with _INJECTOR_LOCK:
+        clear()
+        return configure(config, bind_in_runtime=bind_in_runtime)
+
+
+def is_configured() -> bool:
+    """Return true if an injector is already configured."""
+    with _INJECTOR_LOCK:
+        return _INJECTOR is not None
+
+
+def clear() -> None:
+    """Clear an existing injector if present."""
+    global _INJECTOR
+
+    with _INJECTOR_LOCK:
+        if _INJECTOR is None:
+            return
+
+        _INJECTOR = None
+        logger.debug('Cleared an injector')
+
+
+def instance(cls: T) -> T:
```

Comment:
@wisepotato why was this (and similar functions, e.g. `attr`) annotated to return the same type of argument as the that was given as argument?

I would expect to give there a class (type) as an argument and get a class instance in return, so in my opinion, it should look like this:
```python
def instance(cls: Type[T]) -> T:
    ...
```

This was a version I originally added in .pyi: https://github.com/ivankorobkov/python-inject/pull/21/files#diff-ebdc46604c1a2a1ca81ca20331a18fe5L26

Reply:
Hi, I think, it's just a typo.

Reply:
Does this differ in Python? I did not know about `Type[T]` and `T`, any reference on that?